ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -67,38 +69,43 @@ final class InMemoryLedgerReaderWriter private (
           .getOrElse(StartIndex),
         RangeSource((startExclusive, endInclusive) =>
           Source.fromIterator(() => {
-            val entries = state.readLog(_.zipWithIndex.slice(startExclusive + 1, endInclusive + 1))
-            entries.iterator.map { case (entry, index) => index -> entry }
+            Timed.value(
+              Metrics.readLog,
+              state
+                .readLog(_.view.zipWithIndex.slice(startExclusive + 1, endInclusive + 1).map {
+                  case (entry, index) => index -> entry
+                })
+                .iterator)
           }))
       )
       .map { case (_, updates) => updates }
-}
 
-class InMemoryLedgerStateAccess(currentState: InMemoryState)(
-    implicit executionContext: ExecutionContext
-) extends LedgerStateAccess[Index] {
+  object InMemoryLedgerStateAccess extends LedgerStateAccess[Index] {
+    override def inTransaction[T](body: LedgerStateOperations[Index] => Future[T]): Future[T] =
+      state.write { (log, state) =>
+        body(new InMemoryLedgerStateOperations(log, state))
+      }
+  }
 
-  override def inTransaction[T](body: LedgerStateOperations[Index] => Future[T]): Future[T] =
-    currentState.write { (log, state) =>
-      body(new InMemoryLedgerStateOperations(log, state))
-    }
-}
+  private final class InMemoryLedgerStateOperations(
+      log: InMemoryState.MutableLog,
+      state: InMemoryState.MutableState,
+  ) extends BatchingLedgerStateOperations[Index] {
+    override def readState(keys: Seq[Key]): Future[Seq[Option[Value]]] =
+      Future.successful(keys.map(state.get))
 
-private class InMemoryLedgerStateOperations(
-    log: InMemoryState.MutableLog,
-    state: InMemoryState.MutableState,
-)(implicit executionContext: ExecutionContext)
```

Comment:
Uh, well done.

Reply:
It's just indentation; moved it inside `InMemoryLedgerReaderWriter`.